spec/git_spec.rb

```diff
@@ -154,175 +154,202 @@
       expect(@clone).to be_bare
     end
-    
+
     after(:each) do
       remove_temp_repo(@local)
     end
   end
-  
+
   context "patching" do
     before(:each) do
       @temp_repo_path = create_temp_repo(TEST_REPO_PATH)
       @repo = Repo.new(@temp_repo_path)
     end
-    
+
     it "applies a patch from a String" do
       patch = fixture('postpatriarchialist.patch')
       result = @repo.git.apply_patch(patch)
       expect(result).to have(1).changed_file
       expect(result.first).to match /postpatriarchialist.txt/
     end
-    
+
     it "applies a patch from a file" do
       patch = File.join(File.dirname(__FILE__), 'fixtures', 'postpatriarchialist.patch')
       result = @repo.git.apply_file(patch)
       expect(result).to have(1).changed_file
       expect(result.first).to match /postpatriarchialist.txt/
     end
-    
+
     after(:each) do
       remove_temp_repo(@temp_repo_path)
     end
-    
+
   end
-  
-  describe "push and pull" do 
+
+  describe "push, pull and fetch" do
     before(:each) do
       @temp_repo_path = create_temp_repo(TEST_REPO_PATH)
       @remote = Repo.new(@temp_repo_path)
       @local = @remote.git.clone(@remote.path, get_new_temp_repo_path)
     end
-    
+
     context "pulling from another repository" do
       before(:each) do
         File.open(File.join(@remote.path, "materialist.txt"), 'a') {|file| file.write("\n Beautiful materialist.") }
         @remote.add("materialist.txt")
         @commit = @remote.commit("Making a change for pushing and pulling specs")
       end
-    
+
       it "pulls commits from a local clone" do
         expect(@local.commits).to have(8).commits
         @local.git.pull("origin", "master")
         expect(@local.commits).to have(9).commits
       end
-      
+
       it "pulls commits from a local clone with rebase" do
         expect(@local.commits).to have(8).commits
         @local.git.pull(nil, nil, rebase: true)
         expect(@local.commits).to have(9).commits
       end
-    
+
       it "pulls commits from a local clone with credentials" do
         expect(@local.commits).to have(8).commits
         @local.git.pull(nil, nil, username: 'rspec', password: 'Hahmeid7')
         expect(@local.commits).to have(9).commits
       end
-    
+
+    end
+
+    context "fetching from another repository" do
+      before(:each) do
+        @remote.create_branch('a_new_branch')
+        @remote.checkout('a_new_branch')
+        File.open(File.join(@remote.path, "materialist.txt"), 'a') {|file| file.write("\n Beautiful materialist.") }
+        @remote.add("materialist.txt")
+        @commit = @remote.commit("Making a change for fetching specs")
+        @remote.git.tag('a_new_tag')
+      end
+
+      it "fetches from a local clone" do
+        expect(@local.branches).to have(1).branches
+        expect(@local.tags).to have(0).tags
+        @local.git.fetch(@remote.path, refspecs: "+refs/*:refs/*")
+        expect(@local.branches).to have(3).branches
```

Comment:
So, the remote has two branches, master and alternative. On a clone, the local repo only actually has one branch, master. There is a branch in the local copy of the remote repo, refs/remotes/origin/alternative. Until the branch, alternative, is checked out, it doesn't exist in the local repo. When you do you'll get the message...
```
Branch alternative set up to track remote branch alternative from origin.
Switched to a new branch 'alternative'
``` 
and now there will be 2 branches in the local repo.

However, when I set the refspec to "+refs/*:refs/*" then I take all branches from the remote repo and force them into the local repo instead of the normal fetch behavior of putting them in refs/remotes, which would be a refspec of "+refs/*:refs/remotes/*". So, that is why the branch alternative also becomes available locally as well as the a_new_branch branch.

Does that make sense?

I could also *not* create a new branch, and just use alternative and expect 2. Regardless, I should add a comment here to be clear about what's happening.